Add/email editor sending transactional emails

costasovo opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #55674, Closes #55801.

This PR adds the initial implementation of replacing Woo transactional emails with emails created in the email block editor.

I've provided more details in a GH discussion: https://github.com/woocommerce/woocommerce/discussions/55965

**Do not merge immediately. I would like to get feedback on the approach**

There will be follow-up tasks. Also mentioned in  https://github.com/woocommerce/woocommerce/discussions/55965

![Screenshot 2025-02-28 at 9 25 55](https://github.com/user-attachments/assets/9cf9b99a-88ea-48f2-83d0-024656dad51b)
![Screenshot 2025-02-28 at 9 30 50](https://github.com/user-attachments/assets/748a33ec-a40f-4088-b9cb-3dbcd0194fb9)

### Key Changes

- Added simple placeholder block for WooCommerce content
- Transactional emails that are associated with an email created in the block editor are replaced by the block based emails  during the sending
- The WooCommerce content placeholder is replaced by a content captured from the associated WC_Email during the 
sending
- Currently supported personalized tags are also processed and replaced in the sent email

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the `block_email_editor` experimental flag (in DB, set `woocommerce_feature_block_email_editor_enabled` option to yes).
2. Navigate to WP Admin > Woo Emails
3. Create or edit an email template
4. Use an id of WC_Email you want to replace as the email title/campaign name. This is editable in the component in the center of the header in the editor. The list of WC_Emails can be found [here](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/includes/emails) and you can grab the id from the constructor ([example](https://github.com/woocommerce/woocommerce/blob/65be92caa43dcad3c61b2b1319a2116e748db35b/plugins/woocommerce/includes/emails/class-wc-email-customer-processing-order.php#L30)). This is a temporary way of setting the association.
5. Do the action that triggers the email. For example, for `customer_processing_order` you need to create a new order.
6. You should receive the email created in the block editor and it should contain the main content from the WC_Email. Note: you can use [WP Mail Logging plugin](https://wordpress.org/plugins/wp-mail-logging/) to capture the emails

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

costasovo committed:
Add placeholder block for Woo generated content

This commit adds very basic block that act as a placeholder for
the Woo generated content for the transactional emails (order details, shipping address, etc.).

It is registered only in the email editor via JS so that the block is not globally available.
The idea is that we will extend this block to preview Woo Content in the editor per email type.

costasovo committed:
Add Woo Content placeholder block to the email content pattern

This ensures that when creating a new email with the pattern it
contains the Woo Content block.
The block is locked so that a user can't remove it.

costasovo committed:
Use block email for transactional email when defined.

When the feature is active and there is an email created in the block
editor and associated with the transactional email, we render the block email
using MailPoet\EmailEditor\Engine\Renderer\Renderer.
We grab the main content of Woo email and place it instead of the placeholder
generated by Woo Content block.

To grab only the main content, we need to unregister filters for the email header and footer.
We then restore the filters to be able to correctly send other emails that might not have
an associated email created in the block editor..

costasovo committed:
Add Woo content block render and register it to email renderer

The block will render a placehoder that will be replaced
by the real content during sending

costasovo committed:
Add personalized tags processing when sending transactional email

costasovo committed:
Add more context data for personalization from WC_Email

costasovo committed:
Add changelog

costasovo committed:
Skip block email rendering test if requirements are not fulfilled

The test may run on older WP versions and we need to skip to prevent errors.

triple0t reviewed (commented):
I tested the PR according to the testing notes, and it works well. Well done.

Editor
<img width="890" alt="Screenshot 2025-03-03 at 12 52 25 PM" src="https://github.com/user-attachments/assets/6158096b-c9cf-4063-b0b3-48b40f4b46d1" />

Email content
![Screenshot 2025-03-03 at 2 19 00 PM](https://github.com/user-attachments/assets/71a48412-a017-467e-b54c-67bbc55e9b34)
![Screenshot 2025-03-03 at 2 19 06 PM](https://github.com/user-attachments/assets/7d0be7a3-8d94-4592-a214-1652845f979e)

I also confirmed that the Global styles settings are applied
![Screenshot 2025-03-03 at 2 37 27 PM](https://github.com/user-attachments/assets/c2e657f4-6d72-44d9-b951-841f3659cebc)


I recognize this is an initial draft, so I have some feedback.
I don't have any complaints about the PR approach.

1. Can we register a [select](https://developer.wordpress.org/block-editor/reference-guides/components/select-control/) element (with Woo email templates ID) on the email settings sidebar and allow users to pick the template instead of using the email title/campaign name field to indicate the WooCommerce transaction email template?
2. We can simplify [this section](https://github.com/woocommerce/woocommerce/blob/4325d84486a58a3ab2930f057fccb335666a4538/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L72-L82) by moving the `block_email_renderer->register_hooks` method to an `init` function.
3. We need to create an issue for adding `Order` and `User` related personalization tags.
4. When the block email editor is enabled for a transactional email, I suggested we remove [these other string contents](https://github.com/user-attachments/assets/9f53d56f-f37b-421c-83dc-087e3ecf9b3f) from the email template and pre-populate them on the block email editor content. Note: We don't need to implement the suggestion in this PR.

costasovo committed:
Move BlockEmailRenderer filter registration to init

The init method is called automatically when the service is
created in the container. We can use this to save some lines of code.

costasovo committed:
Add select with email types to the block email editor sidebar.

This commit allows users to select which email type will be overridden
by the block-based email edited in the editor.
The association is currently saved in the post.slug ensuring that
There will be only one block email per email type.

costasovo commented:
@triple0t Thanks for the in-depth review and the suggestions.

> Can we register a [select](https://developer.wordpress.org/block-editor/reference-guides/components/select-control/) element (with Woo email templates ID) on the email settings sidebar and allow users to pick the template instead of using the email title/campaign name field to indicate the WooCommerce transaction email template?

I added it in [d34d74c](https://github.com/woocommerce/woocommerce/pull/55940/commits/d34d74cbfeb6b75919c61db40bbbd9cfc360a475). The value is saved in `wp_post` in the `slug` column. This select component is a temporary solution. I assume that in the end, we end up with something like a pre-generated set of emails and we won't allow users to switch. But it is handy in the meantime.

> We can simplify [this section](https://github.com/woocommerce/woocommerce/blob/4325d84486a58a3ab2930f057fccb335666a4538/plugins/woocommerce/src/Internal/EmailEditor/Integration.php#L72-L82) by moving the block_email_renderer->register_hooks method to an init function.

Done in [0125df8](https://github.com/woocommerce/woocommerce/pull/55940/commits/0125df85402c39024910b268fa114d9f5aeba74a)

> We need to create an issue for adding Order and User related personalization tags.

I created a draft for the issue https://github.com/orgs/woocommerce/projects/297/views/1?pane=issue&itemId=100275855

> When the block email editor is enabled for a transactional email, I suggested we remove [these other string contents](https://github.com/user-attachments/assets/9f53d56f-f37b-421c-83dc-087e3ecf9b3f) from the email template and pre-populate them on the block email editor content. Note: We don't need to implement the suggestion in this PR.

At this point I was looking for a general solution that would work also for the third party emails. I think that in the next steps we will focus on the core's transactional emails one by one. A possible approach for removing the unwanted parts of the content could be adding a condition to the current template and skipping it (something similar [to this](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/emails/customer-processing-order.php#L42-L49)). I would do that in a follow-up task. 

I would like to get this PR merged. There are a few follow-up tasks that we could work on. But I think we should first merge the [PR with namespace refactor](https://github.com/woocommerce/woocommerce/pull/55938) to avoid conflicts.

triple0t commented:
>I added it in [d34d74c](https://github.com/woocommerce/woocommerce/pull/55940/commits/d34d74cbfeb6b75919c61db40bbbd9cfc360a475). The value is saved in wp_post in the slug column. This select component is a temporary solution. I assume that in the end, we end up with something like a pre-generated set of emails and we won't allow users to switch. But it is handy in the meantime.

You might need to update the [query](https://github.com/woocommerce/woocommerce/blob/d34d74cbfeb6b75919c61db40bbbd9cfc360a475/plugins/woocommerce/src/Internal/EmailEditor/BlockEmailRenderer.php#L91-L96) here. The Email editor template is not used when selecting the transaction email from the select component.

> I created a draft for the issue https://github.com/orgs/woocommerce/projects/297/views/1?pane=issue&itemId=100275855
> 
> At this point I was looking for a general solution that would work also for the third party emails. I think that in the next steps we will focus on the core's transactional emails one by one. A possible approach for removing the unwanted parts of the content could be adding a condition to the current template and skipping it (something similar [to this](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/emails/customer-processing-order.php#L42-L49)). I would do that in a follow-up task.

Thank you.

> I would like to get this PR merged. There are a few follow-up tasks that we could work on. But I think we should first merge the https://github.com/woocommerce/woocommerce/pull/55938 to avoid conflicts.

Merged. Please rebase.

costasovo committed:
Update email editor code namespaces and prefixes after rebase

costasovo committed:
Update query for block based emails to use slug aka post_name

costasovo committed:
When there is a block-based email, switch the plain email type to HTML

triple0t reviewed (approved):
> I ran into a case when I defied block based email that is set as plain and it sent HTML code as plain/text. I've addressed that in [455c29f](https://github.com/woocommerce/woocommerce/pull/55940/commits/455c29f91e624ca668bb19e2774757d5ebc55f29)

Instead of hardcoding it [here](https://github.com/woocommerce/woocommerce/blob/455c29f91e624ca668bb19e2774757d5ebc55f29/plugins/woocommerce/includes/emails/class-wc-email.php#L721), I would prefer we respect user preference and send `plain/text` when requested and `HTML` when requested (we can default to `HTML`).

The Block editor renderer [returns](https://github.com/woocommerce/woocommerce/blob/455c29f91e624ca668bb19e2774757d5ebc55f29/plugins/woocommerce/src/Internal/EmailEditor/BlockEmailRenderer.php#L75) both HTML and TEXT. We can pass the user preference across and [return](https://github.com/woocommerce/woocommerce/blob/455c29f91e624ca668bb19e2774757d5ebc55f29/plugins/woocommerce/src/Internal/EmailEditor/BlockEmailRenderer.php#L76-L78) the requested email data.

I'm approving the PR. **Thank you, and well done**.
You can merge to trunk when you're ready.

costasovo commented:
> Instead of hardcoding it [here](https://github.com/woocommerce/woocommerce/blob/455c29f91e624ca668bb19e2774757d5ebc55f29/plugins/woocommerce/includes/emails/class-wc-email.php#L721), I would prefer we respect user preference and send plain/text when requested and HTML when requested (we can default to HTML).

@triple0t Great point! Thinking about it, I don't remember seeing a plain text option in the designs with the block editor. I as for now I assume that block based emails will always be sent as HTML. Also, this seems to be easy to change, so I'll leave it as is for now.

Thank you for the thorough review and valuable remarks! Much appreciated!
